### PR TITLE
Simple TLSv1, take 2

### PR DESCRIPTION
**Pull request by freerobby**
This builds on @alieander's work in https://github.com/wistia/nsq-ruby/pull/25, adding some specs and building toward the interface we agreed on at the end of that PR. Please continue discussion in this PR; I will close the other one.

### CODE REVIEW

File: lib/nsq/connection.rb [hunk 1]
```diff
@@ -32,6 +32,7 @@ def initialize(opts = {})
       @msg_timeout = opts[:msg_timeout] || 60_000 # 60s
       @max_in_flight = opts[:max_in_flight] || 1
       @ssl_context = opts[:ssl_context]
+      @tls_v1 = !!opts[:tls_v1]
       validate_ssl_context! if @ssl_context
```
Comment (alieander) [inline lib/nsq/connection.rb]: If you provide:
` { tls_v1: false, ssl_context: { key: _, cert: _ } }` TLS will not be used but the ssl_context will still be validated.
Given the ssl_context does not validate, errors will still be raised which may be confusing when you have `tls_v1: false`

Comment (freerobby) [inline lib/nsq/connection.rb]: @alieander Good point. The weird `{tls_v1: false, ssl_context: {...}}` case strikes again! I'm thinking maybe we skip validation there and throw some warning to the console that they supplied an ssl context but disabled tls. Thoughts on that?

File: README.md [hunk 2]
```diff
@@ -70,6 +70,7 @@ The Nsq::Producer constructor takes the following options:
 | `nsqd`        | Host and port of the nsqd instance     | '127.0.0.1:4150'   |
 | `nsqlookupd`  | Use lookupd to automatically discover nsqds |               |
 | `ssl_context` | Optional keys and certificates for TLS connections |        |
+| `tls_v1`      | Flag for tls v1 connections (default false) |               |
```
Comment (bschwartz) [inline README.md]: Small nit: want to throw the `(default false)` in the third column?

Comment (bschwartz) [inline README.md]: Also, should this go above `ssl_context`?

**Comment by bschwartz** [comment]
Given that this is a breaking change from the current version (in that folks who have code that sets `ssl_context` will no longer connect via TLS, since `tls_v1` defaults to false), should we also change `ssl_context` to be called `tls_options` to make it more consistent with other client libraries?

Also, mind adding the start of a `CHANGELOG`? Figure that will be helpful for folks who will upgrade.

File: lib/nsq/connection.rb (hunk 1 above)
Comment (alieander) [inline lib/nsq/connection.rb]: Warning sounds like a nice thing to do in that somewhat odd case.
Maybe the `@tls_v1` check and `warn` could go in the `validate_ssl_context!` method like so:

```
def validate_ssl_context!
  unless @tls_v1
    warn 'tls_v1 is false but a context is set: validation skipped'
    return
  end
  ...
end
```

Not 100% sure that is the correct place for it but thinking out loud.

**Comment by mschneider82** [comment]
I would keep the `ssl_context` value to make sure the `warn` is triggered if there is a existing client certificate validation configuration. The boolean `tls_v1` matches the [nsq tcp specs](http://nsq.io/clients/tcp_protocol_spec.html)

File: lib/nsq/connection.rb (hunk 1 above)
Comment (freerobby) [inline lib/nsq/connection.rb]: @alieander Resolved in https://github.com/wistia/nsq-ruby/pull/26/commits/5caa7bd53dfea57db6ab86f769c46db807de7deb

File: README.md (hunk 2 above)
Comment (freerobby) [inline README.md]: Good points, both resolved in https://github.com/wistia/nsq-ruby/pull/26/commits/923dd7b72e489e4d554e48f2820dd264b06594c6

**Comment by freerobby** [comment]
@bschwartz I created a changelog and renamed ssl_context to tls_options.

@mschneider82 We borrowed the tls_options name from pynsq. But to mitigate the point you rase, I do look for ssl_context and use it if tls_options is not provided, so this should catch both cases. Applicable tests are also now run against both usages.

Thanks everyone! I realize this ballooned a bit but think we got it to a good place. Let me know if you have any more suggestions, else I'll merge and release tomorrow.

**Comment by alieander** [comment]
Looks great! Thanks for taking the time to sort everything out!

File: README.md [hunk 3]
```diff
@@ -92,19 +93,28 @@ producer = Nsq::Producer.new(
...
 
-If you need to connect using SSL/TLS
+If you need to connect using SSL/TLS Authentication via `tls_options`
 
 ```Ruby
 producer = Nsq::Producer.new(
   nsqlookupd: ['1.2.3.4:4161', '6.7.8.9:4161'],
   topic: 'topic-of-great-esteem',
-  ssl_context: {
+  tls_options: {
```
Comment (bschwartz) [inline README.md]: I think we should also encourage putting `tls_v1: true` here too, right?

File: README.md [hunk 4]
```diff
@@ -172,7 +182,7 @@ producers when you're done with them.
 | `max_in_flight`      | Max number of messages for this consumer to have in flight at a time   | 1 |
 | `discovery_interval` | Seconds between queue discovery via nsqlookupd    | 60.0   |
 | `msg_timeout`        | Milliseconds before nsqd will timeout a message   | 60000  |
-| `ssl_context`        | Optional keys and certificates for TLS connections |       |
+| `tls_options`        | Optional keys and certificates for TLS connections |       |
```
Comment (bschwartz) [inline README.md]: Shall we add `tls_v1` here too?

File: README.md [hunk 5]
```diff
@@ -185,7 +195,7 @@ consumer = Nsq::Consumer.new(
   max_in_flight: 100,
   discovery_interval: 30,
   msg_timeout: 120_000,
-  ssl_context: {
+  tls_options: {
```
Comment (bschwartz) [inline README.md]: Again, should we put `tls_v1: true` here?

File: lib/nsq/connection.rb [hunk 6]
```diff
@@ -31,8 +31,24 @@ def initialize(opts = {})
...
       @max_in_flight = opts[:max_in_flight] || 1
-      @ssl_context = opts[:ssl_context]
-      validate_ssl_context! if @ssl_context
+      @tls_options = opts[:tls_options]
+      if opts[:ssl_context]
+        if @tls_options
+          warn 'ssl_context and tls_options both set. Using tls_options. Ignoring ssl_context.'
+        else
+          @tls_options = opts[:ssl_context]
+          warn 'ssl_context will be deprecated nsq-ruby version 3. Please use tls_options instead.'
```
Comment (bschwartz) [inline lib/nsq/connection.rb]: Good warnings!

File: spec/lib/nsq/tls_connection_spec.rb [hunk 7]
```diff
@@ -50,4 +78,63 @@ def message_count(topic)
...
+          tls_v1: true
+        }
+        params[tls_options_key] = tls_options_fixture
+
+        conn = Nsq::Connection.new(params)
+        expect(conn.instance_variable_get(:@socket)).
+          to be_an_instance_of(OpenSSL::SSL::SSLSocket)
+        conn.close
+      end
+      it 'tls not used when tls_v1 is false and @tls_options provided' do
```
Comment (bschwartz) [inline spec/lib/nsq/tls_connection_spec.rb]: w00t

File: README.md (hunk 3 above)
Comment (bschwartz) [inline README.md]: Rather, we need that there otherwise this won't work ;)

**Comment by bschwartz** [comment]
@freerobby Nice! Just left some comments on the README. Code looks great. Thanks for sorting all this out -- I think it all makes way more sense and the naming is good.

File: README.md (hunk 5 above)
Comment (freerobby) [inline README.md]: Yep.

File: README.md (hunk 4 above)
Comment (freerobby) [inline README.md]: Yep.

**Comment by freerobby** [comment]
@bschwartz Thanks, agree with all your comments, just pushed changes accordingly.

**Comment by bschwartz** [comment]
Nice! 🚀!

**Comment by freerobby** [comment]
Merged and released. Thank you @alieander for adding this!

**Pull request merged**